File: core/src/identity/keystore_os.rs
//! OS keychain-backed `KeyStore` implementation.
//!
//! Wraps `keyring::Entry::new("xyz.mnemonik.identity", "default")` per
//! Decision 1.  Maps to macOS Keychain, Linux Secret Service (via D-Bus), and
//! Windows Credential Manager.
//!
//! The value stored in the keychain is the compact JSON string:
//!   `{"secret":[...64 bytes...],"pubkey_base58":"..."}`
//! — byte-identical to the legacy file-fallback format per Decision 13.
use keyring::Entry;

use crate::identity::keystore::{KeyStore, KeystoreEntry, KeystoreError};

/// Service and account identifiers for the keychain entry (Decision 1).
const SERVICE: &str = "xyz.mnemonik.identity";
const ACCOUNT: &str = "default";

pub struct OsKeyStore;

impl OsKeyStore {
    pub fn new() -> Self {
        Self
    }

    fn entry(&self) -> Result<Entry, KeystoreError> {
        Entry::new(SERVICE, ACCOUNT).map_err(|e| KeystoreError::Backend(anyhow::anyhow!("{e}")))
    }
}

impl Default for OsKeyStore {
    fn default() -> Self {
        Self::new()
    }
}

impl KeyStore for OsKeyStore {
    /// Retrieve the entry from the OS keychain.
    ///
    /// Returns `Ok(None)` when no entry has been stored yet
    /// (`keyring::Error::NoEntry`).
    fn get(&self) -> Result<Option<KeystoreEntry>, KeystoreError> {
        let entry = self.entry()?;
        match entry.get_password() {
            Ok(json) => {
                let ke: KeystoreEntry = serde_json::from_str(&json)?;
                Ok(Some(ke))
            }
            Err(keyring::Error::NoEntry) => Ok(None),
            Err(keyring::Error::NoStorageAccess(_)) => Err(KeystoreError::PlatformUnavailable {
                reason: "no storage access".to_string(),
            }),
            Err(e) => Err(KeystoreError::Backend(anyhow::anyhow!("{e}"))),
        }
    }

    /// Persist the entry in the OS keychain.
    ///
    /// Serializes to compact JSON with key order `secret` → `pubkey_base58`
    /// (Decision 13) and stores the UTF-8 string.
    fn set(&self, entry: &KeystoreEntry) -> Result<(), KeystoreError> {
        let json = serde_json::to_string(entry)?;
        let ke = self.entry()?;
        match ke.set_password(&json) {
            Ok(()) => Ok(()),
            Err(keyring::Error::NoStorageAccess(e)) => Err(KeystoreError::PlatformUnavailable {
                reason: e.to_string(),
            }),
            Err(e) => Err(KeystoreError::Backend(anyhow::anyhow!("{e}"))),
        }
    }

    /// Delete the keychain entry.  Idempotent — `NoEntry` is treated as success.
    fn remove(&self) -> Result<(), KeystoreError> {
        let entry = self.entry()?;
        match entry.delete_credential() {
            Ok(()) => Ok(()),
            Err(keyring::Error::NoEntry) => Ok(()),
            Err(keyring::Error::NoStorageAccess(e)) => Err(KeystoreError::PlatformUnavailable {
                reason: e.to_string(),
            }),
            Err(e) => Err(KeystoreError::Backend(anyhow::anyhow!("{e}"))),
        }
    }

    /// Lazy probe — does NOT trigger OS keychain unlock prompts (Decision 3).
    ///
    /// Only checks whether the platform backend can construct an `Entry`
    /// descriptor.  Returns `Ok(false)` on `NoStorageAccess` (keychain
    /// subsystem absent — headless Docker, SSH without keyring forwarding,
    /// etc.).  Any other construction error propagates as
    /// `KeystoreError::PlatformUnavailable`.  Does NOT call `get_password()`,
    /// `set_password()`, or `delete_password()`.
    fn available(&self) -> Result<bool, KeystoreError> {
        match Entry::new(SERVICE, ACCOUNT) {
            Ok(_) => Ok(true),
            Err(keyring::Error::NoStorageAccess(_)) => Ok(false),
            Err(e) => Err(KeystoreError::PlatformUnavailable {
                reason: e.to_string(),
            }),
        }
    }

    fn name(&self) -> &'static str {
        "os"
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Real keychain round-trip test.
    ///
    /// Skipped by default so `cargo test` does not touch the OS keychain on CI
    /// without explicit opt-in.  Wave 3 Task 9 exercises this in a
    /// Linux + gnome-keyring container via:
    ///
    ///   cargo test -p mnemonic-core -- --ignored identity::keystore_os
    ///
    #[test]
    #[ignore]
    fn roundtrip_real_keychain() {
        let store = OsKeyStore::new();

        // Clean slate.
        let _ = store.remove();

        // set → get
        let mut secret = [0u8; 64];
        for (i, b) in secret.iter_mut().enumerate() {
            *b = i as u8;
        }
        let entry = KeystoreEntry {
            secret,
            pubkey_base58: "RealKeychainTest".to_string(),
        };
        store.set(&entry).unwrap();

        let got = store.get().unwrap().expect("entry should exist after set");
        assert_eq!(got.secret, secret);
        assert_eq!(got.pubkey_base58, "RealKeychainTest");

        // remove → get == None
        store.remove().unwrap();
        let after_remove = store.get().unwrap();
        assert!(after_remove.is_none(), "entry should be None after remove");
    }
}

File: packages/cli/src/identity/keystore-os.ts
/**
 * OsKeyStore — wraps @napi-rs/keyring to store the Ed25519 keypair in the
 * platform credential store (macOS Keychain, Windows Credential Manager,
 * Linux libsecret/D-Bus).
 *
 * Fixed service/account names so the entry is always addressable:
 *   service  = "xyz.mnemonik.identity"
 *   account  = "default"
 *
 * SECURITY: `secret` bytes are NEVER logged, included in error messages, or
 * placed in Error.cause in any retrievable form.
 */

import { execFileSync } from "node:child_process";

import { Entry } from "@napi-rs/keyring";
import {
  type KeyStore,
  type KeystoreEntry,
  KeystoreError,
  canonicalEntryJson,
} from "./keystore.js";

const SERVICE = "xyz.mnemonik.identity";
const ACCOUNT = "default";

/**
 * macOS-only: write the keychain entry via `security add-generic-password
 * -U -A` so the resulting item's ACL is `kSecACLAuthorizationAny` (allow any
 * application to access without warning). See `set()` doc for why.
 *
 * Args:
 *   -U  Update an existing item if present (idempotent)
 *   -A  Allow any application to access this item without warning
 *   -s  Service name (Decision 1)
 *   -a  Account name (Decision 1)
 *   -w  Password / value (the JSON-encoded `KeystoreEntry`)
 *
 * Argv exposure: the JSON (including the 64-byte secret) appears in `argv`
 * for the duration of the subprocess (~50ms). Documented trade-off; see
 * the Rust mirror in `core/src/identity/keystore_os.rs`.
 */
function setPermissiveMacos(json: string): void {
  try {
    execFileSync(
      "/usr/bin/security",
      [
        "add-generic-password",
        "-U",
        "-A",
        "-s",
        SERVICE,
        "-a",
        ACCOUNT,
        "-w",
        json,
      ],
      { stdio: ["ignore", "ignore", "pipe"] },
    );
  } catch (err) {
    // Do NOT propagate the raw error message — it may contain the json on
    // some shell-error paths. Surface a generic kind only.
    throw new KeystoreError("other", `OsKeyStore.set (macOS shell-out) failed`);
  }
}

/** Classify a thrown error from @napi-rs/keyring into a KeystoreErrorKind. */
function classifyKeychainError(
  err: unknown,
): "platform-unavailable" | "locked" | "other" {
  const msg =
    err instanceof Error
      ? err.message.toLowerCase()
      : String(err).toLowerCase();
  if (
    msg.includes("platform") ||
    msg.includes("no such service") ||
    msg.includes("service") ||
    msg.includes("dbus") ||
    msg.includes("d-bus") ||
    msg.includes("secret service") ||
    msg.includes("not available") ||
    msg.includes("unavailable")
  ) {
    return "platform-unavailable";
  }
  if (
    msg.includes("locked") ||
    msg.includes("denied") ||
    msg.includes("user interaction not allowed")
  ) {
    return "locked";
  }
  return "other";
}

export class OsKeyStore implements KeyStore {
  public readonly name = "os" as const;

  /** Get the stored entry, or null if absent. */
  async get(): Promise<KeystoreEntry | null> {
    let raw: string | null;
    try {
      const entry = new Entry(SERVICE, ACCOUNT);
      raw = entry.getPassword();
    } catch (err) {
      const kind = classifyKeychainError(err);
      // "No matching entry" semantics — treat as absent.
      const msg =
        err instanceof Error
          ? err.message.toLowerCase()
          : String(err).toLowerCase();
      if (msg.includes("no matching") || msg.includes("not found")) {
        return null;
      }
      throw new KeystoreError(
        kind,
        // Do NOT include raw error message text that might contain key material.
        `OsKeyStore.get failed (kind=${kind})`,
        // Omit cause entirely — it may contain the raw JSON with secret bytes.
      );
    }

    if (raw === null) {
      return null;
    }

    let parsed: unknown;
    try {
      parsed = JSON.parse(raw);
    } catch {
      throw new KeystoreError(
        "serde",
        "OsKeyStore: keychain payload is not valid JSON",
      );
    }

    if (
      typeof parsed !== "object" ||
      parsed === null ||
      !("secret" in parsed) ||
      !("pubkey_base58" in parsed)
    ) {
      throw new KeystoreError(
        "serde",
        "OsKeyStore: keychain payload missing required fields",
      );
    }

    const p = parsed as Record<string, unknown>;
    if (!Array.isArray(p["secret"])) {
      throw new KeystoreError(
        "serde",
        "OsKeyStore: 'secret' field is not an array",
      );
    }
    if (typeof p["pubkey_base58"] !== "string") {
      throw new KeystoreError(
        "serde",
        "OsKeyStore: 'pubkey_base58' field is not a string",
      );
    }

    // Normalise to number[] — the stored bytes may come back as a Buffer/Uint8Array
    // after a round-trip through the keychain on some platforms.
    const secretArr: number[] = Array.from(p["secret"] as ArrayLike<number>);

    return {
      secret: secretArr,
      pubkey_base58: p["pubkey_base58"] as string,
    };
  }

  /** Persist the entry in the OS keychain.
   *
   *  **macOS**: shells out to `security add-generic-password -U -A` so the
   *  resulting item's ACL allows any application to read without prompt
   *  (Decision 3 — "no prompts at bootstrap"). `@napi-rs/keyring`'s
   *  `setPassword` creates entries with a restrictive default ACL that
   *  triggers an "Always Allow" GUI prompt for every new caller (e.g. the
   *  Rust mcp-server vs the Node CLI vs a third-party SDK consumer).
   *
   *  **Other platforms**: use `@napi-rs/keyring` directly — Linux Secret
   *  Service and Windows Credential Manager don't have per-item ACLs that
   *  prompt on cross-binary access.
   */
  async set(entry: KeystoreEntry): Promise<void> {
    const json = canonicalEntryJson(entry);

    if (process.platform === "darwin") {
      setPermissiveMacos(json);
      return;
    }

    try {
      const e = new Entry(SERVICE, ACCOUNT);
      e.setPassword(json);
    } catch (err) {
      const kind = classifyKeychainError(err);
      throw new KeystoreError(
        kind,
        `OsKeyStore.set failed (kind=${kind})`,
        // Omit cause — json contains secret bytes.
      );
    }
  }

  /** Remove the entry from the OS keychain. Idempotent. */
  async remove(): Promise<void> {
    try {
      const e = new Entry(SERVICE, ACCOUNT);
      e.deletePassword();
    } catch (err) {
      const msg =
        err instanceof Error
          ? err.message.toLowerCase()
          : String(err).toLowerCase();
      if (msg.includes("no matching") || msg.includes("not found")) {
        return; // already absent — idempotent
      }
      const kind = classifyKeychainError(err);
      throw new KeystoreError(kind, `OsKeyStore.remove failed (kind=${kind})`);
    }
  }

  /**
   * Return true if the OS keychain is available on this platform.
   *
   * Constructs a new Entry and performs a lightweight probe WITHOUT calling
   * getPassword() (which would trigger OS prompts). We rely on the fact that
   * @napi-rs/keyring throws synchronously at construction time on platforms
   * where the backend is completely absent (e.g. headless Linux without D-Bus).
   *
   * If construction succeeds, we assume the backend exists; we cannot know
   * whether it is locked without actually calling getPassword().
   */
  async available(): Promise<boolean> {
    try {
      new Entry(SERVICE, ACCOUNT);
      return true;
    } catch (err) {
      const kind = classifyKeychainError(err);
      if (kind === "platform-unavailable") {
        return false;
      }
      // Unexpected error — re-throw so callers know something went wrong.
      throw new KeystoreError(
        kind,
        `OsKeyStore.available probe failed (kind=${kind})`,
      );
    }
  }
}
